### PR TITLE
Refactor bernstein vazirani

### PR DESCRIPTION
**Pull request by jotterbach**

### CODE REVIEW

File: grove/alpha/bernstein_vazirani/bernstein_vazirani.py [hunk 1]
```diff
@@ -1,171 +1,203 @@
...
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+##############################################################################
+
+"""
+Module for the Bernstein-Vazirani Algorithm.
+For more information, see [Loceff2015]_
+
+.. [Loceff2015] Loceff, M. (2015), `"A Course in Quantum Computing for the Community College"`_, Volume 1, Chapter 18, p 484-541.
+
```
Comment (ampolloreno) [inline grove/alpha/bernstein_vazirani/bernstein_vazirani.py]: Can we honor the line limit?

File: grove/alpha/bernstein_vazirani/bernstein_vazirani.py [hunk 2]
```diff
@@ -1,171 +1,203 @@
...
+    def _create_bv_circuit(self, bit_map):
+        """
+        Implementation of the Bernstein-Vazirani Algorithm.
+
+        Given a list of input qubits and an ancilla bit, all initially in the
+        :math:`\\vert 0\\rangle` state, create a program that can find :math:`\\vec{a}` with one
+        query to the given oracle.
+
+        :param Program oracle_circuit: Program representing unitary application of function
```
Comment (ampolloreno) [inline grove/alpha/bernstein_vazirani/bernstein_vazirani.py]: wrong docstring

File: grove/alpha/bernstein_vazirani/bernstein_vazirani.py [hunk 3]
```diff
@@ -1,171 +1,203 @@
...
-    for i in range(n):
-        if vec_a[i] == 1:
-            p.inst(CNOT(qubits[n - 1 - i], ancilla))
-    return p
-
-
-def bernstein_vazirani(oracle, qubits, ancilla):
+           (\\mathbf{a}\\in\\{0,1\\}^n, b\\in\\{0,1\\})
+    :param String a: a string of 0's and 1's that represents the dot-product partner in :math:`f`
+    :param String b: 0 or 1 as a string representing the bias term in :math:`f`
```
Comment (ampolloreno) [inline grove/alpha/bernstein_vazirani/bernstein_vazirani.py]: The listed reference at the top doesn't talk about a bias term.

File: grove/alpha/bernstein_vazirani/bernstein_vazirani.py [hunk 4]
```diff
@@ -1,171 +1,203 @@
...
-        p.inst(X(ancilla))
-    for i in range(n):
-        if vec_a[i] == 1:
-            p.inst(CNOT(qubits[n - 1 - i], ancilla))
-    return p
-
-
-def bernstein_vazirani(oracle, qubits, ancilla):
+           (\\mathbf{a}\\in\\{0,1\\}^n, b\\in\\{0,1\\})
+    :param String a: a string of 0's and 1's that represents the dot-product partner in :math:`f`
```
Comment (ampolloreno) [inline grove/alpha/bernstein_vazirani/bernstein_vazirani.py]: can we give more descriptive variable names?

File: grove/alpha/bernstein_vazirani/bernstein_vazirani.py [hunk 5]
```diff
@@ -1,171 +1,203 @@
...
-    p = pq.Program()
-
-    # Put ancilla bit into minus state
-    p.inst(X(ancilla), H(ancilla))
+    n_bits = len(a)
+    bit_map = {}
+    for bit_val in range(2 ** n_bits):
+        bit_map[np.binary_repr(bit_val, width=n_bits)] = \
+            str((int(utils.bitwise_dot_product(
+                np.binary_repr(bit_val, width=n_bits), a)) + int(b, 2)) % 2)
```
Comment (ampolloreno) [inline grove/alpha/bernstein_vazirani/bernstein_vazirani.py]: Can we get rid of the backslash?:
PEP8:
"The preferred way of wrapping long lines is by using Python's implied line continuation inside parentheses, brackets and braces. Long lines can be broken over multiple lines by wrapping expressions in parentheses. These should be used in preference to using a backslash for line continuation."

File: grove/alpha/bernstein_vazirani/bernstein_vazirani.py (hunk 3 above)
Comment (ampolloreno) [reply grove/alpha/bernstein_vazirani/bernstein_vazirani.py]: Ah, but the docs do.

File: grove/alpha/bernstein_vazirani/bernstein_vazirani.py [hunk 6]
```diff
@@ -1,171 +1,203 @@
...
+        return full_bv_circuit
+
+    def run(self, cxn, bitstring_map):
+        """
+        Runs the Bernstein-Vazirani algorithm.
+
+        Given a QVM connection, find the :math:`\\mathbf{a}` and :math:`b` corresponding to the
+        function represented by the oracle.
+
+        :param Connection cxn: the QVM connection to use to run the programs
```
Comment (ampolloreno) [inline grove/alpha/bernstein_vazirani/bernstein_vazirani.py]: QVM or QPU?

File: grove/alpha/bernstein_vazirani/bernstein_vazirani.py (hunk 1 above)
Comment (jotterbach) [reply grove/alpha/bernstein_vazirani/bernstein_vazirani.py]: done (I think)

File: grove/alpha/bernstein_vazirani/bernstein_vazirani.py [hunk 7]
```diff
@@ -1,171 +1,203 @@
...
+.. _`"A Course in Quantum Computing for the Community College"`: http://lapastillaroja.net/wp-content/uploads/2016/09/Intro_to_QC_Vol_1_Loceff.pdf
 """
 
 import numpy as np
 import pyquil.quil as pq
-from pyquil.gates import *
-from six.moves import input
+from pyquil.gates import H, X
+from grove.alpha.bernstein_vazirani import utils
+from collections import defaultdict
```
[the remark this answers is not in the document]
Comment (jotterbach) [reply grove/alpha/bernstein_vazirani/bernstein_vazirani.py]: done

File: grove/alpha/bernstein_vazirani/bernstein_vazirani.py (hunk 4 above)
Comment (jotterbach) [reply grove/alpha/bernstein_vazirani/bernstein_vazirani.py]: sure ... tried to stay consistent with the docstrings

File: grove/alpha/bernstein_vazirani/bernstein_vazirani.py (hunk 3 above)
Comment (jotterbach) [reply grove/alpha/bernstein_vazirani/bernstein_vazirani.py]: 👍

File: grove/alpha/bernstein_vazirani/bernstein_vazirani.py (hunk 5 above)
Comment (jotterbach) [reply grove/alpha/bernstein_vazirani/bernstein_vazirani.py]: done

File: grove/alpha/bernstein_vazirani/bernstein_vazirani.py (hunk 2 above)
Comment (jotterbach) [reply grove/alpha/bernstein_vazirani/bernstein_vazirani.py]: good catch!

File: grove/alpha/bernstein_vazirani/bernstein_vazirani.py (hunk 6 above)
Comment (jotterbach) [reply grove/alpha/bernstein_vazirani/bernstein_vazirani.py]: 👍

File: grove/alpha/bernstein_vazirani/utils.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
+from operator import xor
+
+PADDED_BINARY_BIT_STRING = "{0:0{1:0d}b}"
+
+
+def bitwise_dot_product(bs0, bs1):
+    """
+    A helper to calculate the bitwise dot-product between two string representing bit-vectors
```
[the remark this answers is not in the document]
Comment (jotterbach) [reply grove/alpha/bernstein_vazirani/utils.py]: done

File: grove/alpha/bernstein_vazirani/utils.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    :param bs0: String of 0's and 1's representing a number in binary representations
+    :param bs1: String of 0's and 1's representing a number in binary representations
```
[the remark this answers is not in the document]
Comment (jotterbach) [reply grove/alpha/bernstein_vazirani/utils.py]: done

File: grove/alpha/bernstein_vazirani/utils.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    :rtype: String
+    """
+    if len(bs0) != len(bs1):
+        raise ValueError("Bit strings are not of equal length")
+    return str(sum([int(bs0[i]) * int(bs1[i]) for i in range(len(bs0))]) % 2)
+
+
+def bitwise_xor(bs0, bs1):
+    """
+    A helper to calculate the bitwise XOR of two bit string
```
[the remark this answers is not in the document]
Comment (jotterbach) [reply grove/alpha/bernstein_vazirani/utils.py]: doen

File: grove/alpha/bernstein_vazirani/utils.py [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+    :param bs0: String of 0's and 1's representing a number in binary representations
+    :param bs1: String of 0's and 1's representing a number in binary representations
```
[the remark this answers is not in the document]
Comment (jotterbach) [reply grove/alpha/bernstein_vazirani/utils.py]: done

**Pull request merged**